akka-stream/src/main/scala/akka/stream/impl/io/TLSActor.scala

```diff
@@ -414,7 +414,11 @@ import akka.util.ByteString
             handshakeFinished()
             transportInChoppingBlock.putBack(transportInBuffer)
           case _ =>
-            if (transportInBuffer.hasRemaining) doUnwrap(ignoreOutput = false)
+            if (transportInBuffer.hasRemaining)
+              if (userOutBuffer.position() == 0)
+                throw new IllegalStateException("SSLEngine trying to loop NEED_UNWRAP without producing output")
```

Comment:
This would now catch all cases where we loop directly with `@tailrec`. There might be other cases where the looping goes through the pump flip-flopping between `doInbound` and `doOutbound`.

Reply:
perhaps we should include `result.getHandshakeStatus` in the exception message, so we have a bit more to go on when someone encounters this on an SSL connection that should actually be healthy? Or can we somehow be sure it must be in `NEED_UNWRAP`?